File: app/leetcode/No3.lengthOfLongestSubstring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  // 暴力法
  // let len = 0;
  // // 双指针,[i,j]
  // let i = 0, j = 0;
  // while (i < s.length) {
  //   j = i;
  //   const set = new Set();
  //   while (j < s.length && !set.has(s[j])) {
  //     set.add(s[j]);
  //     j++;
  //   }
  //   len = Math.max(len, set.size);
  //   i++;
  // }
  // return len;


  // 滑动窗口，例如字符串(a)bcabc，的最长无重复子串为(abc)abc,a(b)cabc的最长无重复子串为a(bca)bc
  // 我们可以得到：无重复元素的左边界为i,右边界为j，[i,j]为闭区间，则[i+1,j]最起码是无重复的，即以i+1起始的字符串我们不需要一个字符一个字符去尝试（上面的暴力法），只需要从j开始尝试，直到右侧出现了重复字符

  const set = new Set(); // 记录某个字符是否出现过
  let i = 0; // 左指针
  let j = -1; // 右指针，初始值为 -1，相当于我们在字符串的左边界的左侧，还没有开始移动
  const n = s.length;

  let len = 0;

  while (i < n) {
    // 左指针向右移动一格，移除窗口中的一个字符
    if (i !== 0) {
      set.delete(s[i - 1]);
    }
    // 如果窗口中没有重复元素，则不断移动右指针放大窗口
    while (j + 1 < n && !set.has(s[j + 1])) {
      set.add(s[++j]);
    }
    // 窗口[i,j]是无重复元素的最长子串
    len = Math.max(len, j - i + 1);
    i++;
  }
  return len;
};

// console.log(lengthOfLongestSubstring('abcabcbb'));
// console.log(lengthOfLongestSubstring('bbbbb'));
console.log(lengthOfLongestSubstring('pwwkew'));

File: app/leetcode/No3.lengthOfLongestSubstring.go
package main

import "fmt"

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func lengthOfLongestSubstring(s string) int {
	n := len(s)
	if n == 0 {
		return 0
	}
	l := -1
	// 暴力法
	for i := 0; i < n; i++ {
		set := map[byte]struct{}{}
		for j := i; j < n; j++ {
			if _, ok := set[s[j]]; ok {
				break
			} else {
				set[s[j]] = struct{}{}
			}
		}
		l = max(l, len(set))
	}
	return l
}

func lengthOfLongestSubstring2(s string) int {
	// 维护一个滑动窗口，保证窗口内不存在重复元素
	// 对字符串进行遍历，每遍历一个字符就把它加入到窗口之中，同时更新窗口的左边界，保证窗口内没有重复元素
	// 可以看出窗口左边界的维护，依赖于新加入字符上一次出现的位置

	// 滑动窗口: 解题模板
	//外层循环扩展右边界，内层循环扩展左边界
	// for (int l = 0, r = 0 ; r < n ; r++) {
	// 	//当前考虑的元素
	// 	while (l <= r && check()) {//区间[left,right]不符合题意
	//         //扩展左边界
	//     }
	//     //区间[left,right]符合题意，统计相关信息
	// }
	set := map[byte]bool{}
	maxLen := 0
	for left, right := 0, 0; right < len(s); right++ { // 每一轮扩展右边界
		ch := s[right] // right 指向的元素是当前需要考虑的元素
		for left < right && set[ch] {
			delete(set, s[left]) // set 中有元素，则缩短左边界，同时从 set 中移除元素
			left++
		}
		set[ch] = true // 将当前元素加入
		maxLen = max(maxLen, right-left+1)
	}
	return maxLen
}

func lengthOfLongestSubstring3(s string) int {
	mp := map[byte]int{} // 记录每个字符上一次出现的位置
	maxLen := 0
	for l, r := 0, 0; r < len(s); r++ {
		// 上面滑动窗口的优化，左边界不需要一个个向右侧移动，只需要一次性到位，直接跳到下一个不重复的位置就行
		if index, ok := mp[s[r]]; ok && index >= l {
			l = index + 1
		}
		mp[s[r]] = r
		maxLen = max(maxLen, r-l+1)
	}
	return maxLen
}

func main() {
	// s := "abcabcbb"
	// s := "bbbbb"
	s := "pwwkew"
	r := lengthOfLongestSubstring2(s)
	fmt.Println(r)
}
